typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1302 @@
+        | Variant_field l ->
+            Format.fprintf ppf "the field \"%s\" of the variant" l
+        | Memory_address -> Format.fprintf ppf ""
+
+      let to_string (t : t) = Format.asprintf "%a" print t
+
+      (* Yes, compare based on string is bad, but it saves 20 lines of spaghetti
+         code. Also I believe to_string is injective so it might actaully be
+         fine. *)
+      let compare t1 t2 = String.compare (to_string t1) (to_string t2)
+    end
+
+    include T
+    module Map = Map.Make (T)
+  end
+
+  type relation = Ancestor | Descendant
+
+  (* Tree: Each node records the par on all possible execution paths. As a
+     result, trees such as `S -> U` is valid, even though it would be invalid if
+     it was the result of a single path: using a parent shared and a child
+     uniquely is obviously bad. Howerver, it might be the result of "par"ing
+     multiple path: `S` par `N -> U`, which is valid. *)
+
+  (* INVARIANT: children >= parent. For example, having a shared child under a
+     unique parent is nonsense. The invariant is preserved because Usage.par and
+     Usage.seq above is monotone, and UsageTree.par and UsageTree.seq here is
+     node-wise. *)
+  type t = { children : t Projection.Map.t; usage : Usage.t }
+
+  let rec print_children ppf children =
+    Projection.Map.iter
+      (fun proj child ->
+        Format.fprintf ppf "%a = %a," Projection.print proj print child)
+      children
+
+  and print ppf t =
+    Format.fprintf ppf "%a {%a}" Usage.print t.usage print_children t.children
+
+  module Path = struct
+    type t = Projection.t list
+
+    let child (p : t) (a : Projection.t) : t = p @ [ a ]
+
+    (** length of path ignoring memory_address *)
+    let rec real_length = function
+      | [] -> 0
+      | Projection.Memory_address :: xs -> real_length xs
+      | _ :: xs -> real_length xs + 1
+
+    let root : t = []
+    let _print ppf = Format.pp_print_list Projection.print ppf
+
+    let _human_readable ppf t =
+      Format.pp_print_list
+        ~pp_sep:(fun ppf () -> Format.fprintf ppf " which is ")
+        Projection.human_readable ppf t
+  end
+
+  exception
+    MultiUse of {
+      (* the occurrence that's failing forcing *)
+      here : Occurrence.t;
+      (* the other occurrence for reference *)
+      there : Occurrence.t;
+      (* which axis failed to force? *)
+      error : SharedUnique.error;
+      (* descentdant means there is a descendant of here *)
+      there_is_of_here : relation;
+      (* path from the ancestor to the descendant, reversed *)
+      path : Path.t;
+    }
+
+  let leaf usage = { usage; children = Projection.Map.empty }
+  let empty = leaf Usage.Unused
+
+  (* lift par and seq to trees *)
+
+  let rec par t0 t1 =
+    {
+      usage = Usage.par t0.usage t1.usage;
+      children =
+        Projection.Map.merge
+          (fun _proj c0 c1 ->
+            let c0 = Option.value c0 ~default:t0 in
+            let c1 = Option.value c1 ~default:t1 in
```

Comment:
There is some opportunity to short-circuit some cases here. For instance, `t1.usage` < `t0.usage` and `c1` is `None` then we can just return `c0`.

Reply:
`t1.usage < t0.usage` is hard to define properly - I can write a conservative function but it won't detect all cases - for example, it's hard to compare two `MaybeUnique`, as they contain a list of modes.

Besides, it seems that we can factor out some common code out of `par` and `seq` and call it `Usage_tree.map2`, and `seq` doesn't enjoy the same property that you described.